sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -211,6 +215,65 @@ class SQLBuilder(logicalPlan: LogicalPlan, sqlContext: SQLContext) extends Loggi
     )
   }
 
+  private def groupingSetToSQL(
+      plan: Aggregate,
+      expand: Expand,
+      project: Project): String = {
+    require(plan.groupingExpressions.length > 1)
+
+    // The last column of Expand is always grouping ID
+    val gid = expand.output.last
+
+    val groupByAttributes = plan.groupingExpressions.dropRight(1).map(_.asInstanceOf[Attribute])
+    val groupByExprs =
+      project.projectList.drop(project.child.output.length).map(_.asInstanceOf[Alias].child)
+    val groupByAttrMap = AttributeMap(groupByAttributes.zip(groupByExprs))
+    val groupingSQL = groupByExprs.map(_.sql).mkString(", ")
+
+    val groupingSet = expand.projections.map { project =>
+      project.dropRight(1).collect {
+        case attr: Attribute if groupByAttrMap.contains(attr) => groupByAttrMap(attr)
+      }
+    }
+    val groupingSetSQL =
+      "GROUPING SETS(" +
+        groupingSet.map(e => s"(${e.map(_.sql).mkString(", ")})").mkString(", ") + ")"
+
+    val aggExprs = plan.aggregateExpressions.map { case expr =>
+      expr.transformDown {
+        case a @ Alias(child: AttributeReference, name) if child eq gid =>
```

Comment:
seems we get the alias name and use it to build a new `Alias` only? `transformDown` will do this trick for us I think.